rust/etw_helpers/src/event_record.rs

```diff
@@ -0,0 +1,249 @@
+//use core::mem::size_of;
+use std::borrow::Cow;
+use windows::Win32::System::Diagnostics::Etw::{
+    ETW_BUFFER_CONTEXT, EVENT_HEADER, EVENT_HEADER_EXTENDED_DATA_ITEM, EVENT_RECORD,
+};
+
+pub enum EventRecord {
+    Owned(EventRecordOwned),
+    Raw(*const EVENT_RECORD),
+    Serialized(Vec<u8>), // EventRecordSerialized as little endian bytes
+}
+
+#[derive(Clone)]
+struct ExtendedDataItemOwned {
+    ext_type: u16,
+    flags: u16, // Bit 1: linkage, Bit 2-16: unused
+    data: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct EventRecordOwned {
+    event_header: EVENT_HEADER,
+    buffer_context: ETW_BUFFER_CONTEXT,
+    extended_data_items: Vec<ExtendedDataItemOwned>,
+    user_data: Vec<u8>,
+}
+
+// Compatible with TDH
+// #[repr(C)]
+// struct EventRecordSerialized {
+//     Size: u32,
+//     EventHeader: EVENT_HEADER,
+//     BufferContext: ETW_BUFFER_CONTEXT,
+//     // data: [u8; Size - sizeof(self)]
+// }
+
+impl EventRecord {
+    pub fn new(evt: *const EVENT_RECORD) -> Self {
+        // TODO: Check pointers and lengths
+        EventRecord::Raw(evt)
```

Comment:
This results in unverifiable behavior (the caller must now manually track lifetime). We can't do this in a safe method of a public library. Alternatives:

- Mark this method as unsafe, and carefully document what has to happen to avoid trouble with this method, making all users of the EventRecord struct become unsafe and be responsible for managing the lifetime of this object.
- Make a change in the design of this struct so that it can be used safely.

I strongly recommend option 2 because it's really unkind to the users of this library to push the safety issues onto them when we could do better.

Option 2 probably requires splitting EventRecord into two types, something like `EventRecordOwned` and `EventRecordRef<'a>` (where the Ref type has a lifetime).